File: functions/alsa-lib/variants/go/0_alaw_to_s16_4.go
package main

func alawToS16(aVal byte) int32 {
	aVal ^= 0x55
	t := int32(aVal & 0x7f) * 8
	if t < 128 {
		t += 8
	} else {
		seg := (t >> 7) & 0x07
		t = (t & 0x7f) + 0x108
		t <<= seg
	}
	if aVal&0x80 != 0 {
		return t
	}
	return -t
}


func main(){
    alawToS16('a')
}

File: functions/alsa-lib/variants/c/0_alaw_to_s16_4.c
#include <stdint.h>

static int alaw_to_s16_v5(unsigned char a_val)
{
	int t, seg;
	a_val ^= 0x55;
	seg = (a_val >> 4) & 0x07;
	t = 0x10 + (a_val & 0x0F);
	t <<= (seg >= 1) ? (seg + 2) : 4;
	return (a_val & 0x80) ? t : -t;
}


int main(){
    alaw_to_s16_v5('a');
}
